sql/core/src/main/scala/org/apache/spark/sql/Column.scala

```diff
@@ -200,6 +203,7 @@ class Column(val expr: Expression) extends Logging {
     // Wait until the struct is resolved. This will generate a nicer looking alias.
     case struct: CreateNamedStruct => UnresolvedAlias(struct)
 
+    case expr: Expression if existsUnresolvedExtractValue(expr) => UnresolvedAlias(expr)
```

Comment:
Why do we only check `UnresolvedExtractValue` recursively? How about others like `CreateNamedStruct`, `AggregateExpression`, etc.?

Reply:
The reason why not check other `Expression` is `UnresolvedAlias` will introduce `CAST` to assgin name duing analysis if exists type cast implicitly. As I said above 
> Before that df.selectExpr("id1 == null").schema.head.name is (id1 = NULL), after that it changed to (id1 = CAST(NULL AS INT)).

Narrow to checked `UnresolvedExtractValue` is just a tradeoff.

Reply:
won't `UnresolvedExtractValue` have the same problem? e.g. `struct_col.field = NULL`

Reply:
It has the same problem, but `struct_col.field` is much better than `struct_col[field]`.

Other hand, `sql("select id == null")` also has the `CAST` which is different with `selectExpr("select id == null")`. It depends on when to assign name (before or after analysis).